host.go

```diff
@@ -53,81 +97,89 @@ func newRunner(m manifest.ManifestProvider, client *client) (*gpbftRunner, error
 	finalityCertificates := make(chan *certs.FinalityCertificate, 4)
-	_, _ = h.client.certStore.SubscribeForNewCerts(finalityCertificates)
-
-	defer func() {
-		if _err != nil {
-			h.log.Errorf("gpbfthost exiting: %+v", _err)
-		}
-	}()
-
-	messageQueue := h.client.IncomingMessages()
-	for {
+	_, _ = h.certStore.SubscribeForNewCerts(finalityCertificates)
 
-		// prioritise finality certificates and alarm delivery
-		select {
-		case c, ok := <-finalityCertificates:
-			if !ok {
-				finalityCertificates = make(chan *certs.FinalityCertificate, 4)
-				c, _ = h.client.certStore.SubscribeForNewCerts(finalityCertificates)
-			}
-			if err := h.receiveCertificate(c); err != nil {
-				return err
+	h.errgrp.Go(func() (_err error) {
+		defer func() {
+			// TODO: catch panics
+			if _err != nil && h.runningCtx.Err() == nil {
+				h.log.Errorf("exited GPBFT runner early: %+v", _err)
 			}
-			continue
-		case <-h.alertTimer.C:
-			if err := h.participant.ReceiveAlarm(); err != nil {
-				return err
+		}()
+		for h.runningCtx.Err() == nil {
+			// prioritise finality certificates and alarm delivery
+			select {
+			case c, ok := <-finalityCertificates:
+				if !ok {
+					finalityCertificates = make(chan *certs.FinalityCertificate, 4)
+					c, _ = h.certStore.SubscribeForNewCerts(finalityCertificates)
+				}
+				if err := h.receiveCertificate(c); err != nil {
+					return err
+				}
+				continue
+			case <-h.alertTimer.C:
+				if err := h.participant.ReceiveAlarm(); err != nil {
+					return err
+				}
+				continue
+			default:
 			}
-			continue
-		default:
-		}
 
-		// Handle messages, finality certificates, and alarms
-		select {
-		case c, ok := <-finalityCertificates:
-			if !ok {
-				finalityCertificates = make(chan *certs.FinalityCertificate, 4)
-				c, _ = h.client.certStore.SubscribeForNewCerts(finalityCertificates)
-			}
-			if err := h.receiveCertificate(c); err != nil {
-				return err
-			}
-		case <-h.alertTimer.C:
-			if err := h.participant.ReceiveAlarm(); err != nil {
-				return err
-			}
-		case msg, ok := <-messageQueue:
-			if !ok {
-				return xerrors.Errorf("incoming message queue closed")
-			}
-			if err := h.participant.ReceiveMessage(msg); err != nil {
-				return err
+			// Handle messages, finality certificates, and alarms
+			select {
+			case c, ok := <-finalityCertificates:
+				if !ok {
+					finalityCertificates = make(chan *certs.FinalityCertificate, 4)
+					c, _ = h.certStore.SubscribeForNewCerts(finalityCertificates)
+				}
+				if err := h.receiveCertificate(c); err != nil {
+					return err
+				}
+			case <-h.alertTimer.C:
+				if err := h.participant.ReceiveAlarm(); err != nil {
+					return err
+				}
+			case msg, ok := <-messageQueue:
+				if !ok {
+					return xerrors.Errorf("incoming message queue closed")
+				}
+				if err := h.participant.ReceiveMessage(msg); err != nil {
+					// XXX: Log only?
+					h.log.Debugf("error when processing message: %+v", err)
```

Comment:
Is this a hard error?

Reply:
I changed it to a log because it didn't look like it.

Reply:
I'm pretty sure this is not a hard error and it's something we should test.

Reply:
I think checking for PanicErr would be worth it

Reply:
We'll do this in a followup.

Reply:
https://github.com/filecoin-project/go-f3/issues/421